### ISSUE TITLE
Make BPB evaluation fail closed and correct in fused mode

### ISSUE THREAD

**Issue by WhymustIhaveaname** [suning-git/nanoinfra#6]
## Definition

Bits per byte should be computed as:

```text
BPB = sum(per-token cross-entropy in nats for byte-bearing text tokens)
      / (ln(2) * total UTF-8 bytes)
```

Equivalently, if cross-entropy is already measured with log2, BPB is `sum CE_bits / total bytes`.

There are two current correctness problems.

## 1. Missing byte table silently changes the metric

`modalities.text.tokenizer.get_token_bytes()` returns an all-ones table when `token_bytes.pt` is missing. Callers continue to report the result under a `bpb` metric name, but an all-ones denominator computes bits per token, not bits per byte. A warning does not prevent incorrect metrics from entering logs or experiment comparisons. The shipped text exemplar documents that historical values around 5.5 were actually bits/token and later remeasured near 1.24 BPB.

When BPB is requested, a missing or incompatible byte table should fail closed (or omit BPB explicitly), not substitute a different metric under the same name.

## 2. `evaluate_loss_fused()` cannot derive BPB from a scalar mean loss

The fused path currently does approximately:

```python
loss = model.head.loss(hidden, y)       # scalar mean over all valid targets
total_nats += loss * bpb_valid.sum()    # only count byte-bearing targets
```

This is incorrect when zero-byte control tokens are present. Their losses are included in the scalar mean, then that mixed mean is assigned to the byte-bearing tokens. The logits path correctly has per-token losses and sums only `loss_flat[bpb_valid]`.

A minimal two-token reproduction (one zero-byte control target with high NLL and one two-byte text target with low NLL) gives:

```text
evaluate_loss_logits BPB = 0.0000327619
evaluate_loss_fused  BPB = 3.6067706229
```

## Expected behavior

- BPB requests require a real byte table whose length matches the vocabulary; missing/mismatched artifacts fail closed or suppress the metric explicitly.
- logits and fused modes produce equivalent BPB within an appropriate numerical tolerance.
- zero-byte control targets contribute neither NLL nor bytes to BPB.
- tests include nonuniform per-token losses, zero-byte control tokens, ignored targets, and a missing byte-table case.

If the fused head exposes only a scalar mean loss, it does not contain enough information to compute filtered BPB correctly. The API must either expose per-token/sufficient grouped loss sums or declare BPB unsupported in fused mode rather than emit a wrong value.

**Comment by WhymustIhaveaname** [comment]
Resolved in commit f23554c.

Both problems are addressed:

1. **Fail closed on a missing or mismatched byte table.** `get_token_bytes()` now raises `FileNotFoundError` when `token_bytes.pt` is absent, and `ValueError` when the table is not a 1-D integer tensor whose length matches the vocabulary. The all-ones fallback is gone. `modalities/text/evaluator.py` only loads the table when the stream declares `bpb_metric`, so streams that do not report bpb are unaffected.

2. **Fused path no longer derives bpb from a mixed scalar mean.** `evaluate_loss_fused()` now masks the zero-byte targets to `IGNORE_INDEX` and runs `head.loss` a second time, so the mean is taken over the byte-bearing set only. The extra head pass is issued unconditionally because `head.loss` is an FSDP forward method and every rank must call it the same number of times. An empty byte-bearing set contributes 0 nats via `torch.where`, so a genuine NaN on a non-empty set still propagates.

The commit also documents a related artifact defect found at the same time: the `token_bytes.pt` in use before 2026-09-11 overstated 216 of 32768 ids (the byte-fallback ids and merges ending mid-character were counted through U+FFFD), which made every earlier bpb read low by roughly 0.08%. See `exemplars/text_pretrain/provenance.md`.

Tests added: `modalities/tests/test_token_bytes.py` and `core/tests/unit/test_eval_loss_bpb.py`.

Closing.
